### PR TITLE
Fixes #243: Installer should support GPG verification

### PR DESCRIPTION
**Pull request by janosdebugs**

### CODE REVIEW

**Review by cam72cam — APPROVED** [review]

File: src/install-opentofu.ps1 [hunk 1]
```diff
@@ -23,42 +23,68 @@ Path to cosign. (Standalone installation only.)
...
+Path for the GPG installation. (Standalone installation only.)
+.PARAMETER gpgURL
+URL to download the GPG key from. (Standalone installation only.)
+.PARAMETER gpgKeyID
+GPG key ID / fingerprint to expect. (Standalone installation only.)
 .PARAMETER skipVerify
-Skip cosign integrity verification. (Standalone installation only; not recommended.)
-.Parameter skipChangePath
+Skip cosign/GPG integrity verification. (Standalone installation only; not recommended.)
+.PARAMETER skipChangePath
```
Comment (pdecat) [inline src/install-opentofu.ps1:34]: This comment about `skipChangePath` is incorrect:
```suggestion
Skip changing the user/system path to include the OpenTofu path
.PARAMETER skipChangePath
```

See https://github.com/opentofu/get.opentofu.org/blob/ddfdb1e4388db6f29285cace3d7e9cb77de188c7/src/install-opentofu.ps1#L238-L259
and https://github.com/opentofu/get.opentofu.org/blob/ddfdb1e4388db6f29285cace3d7e9cb77de188c7/src/install-opentofu.ps1#L565C49-L565C112

File: src/install-opentofu.ps1 [hunk 2]
```diff
@@ -23,42 +23,68 @@ Path to cosign. (Standalone installation only.)
...
 Skip changing the user/system PATH variable to include OpenTofu.
```
Comment (pdecat) [inline src/install-opentofu.ps1:35]:
```suggestion
Install for all users with elevated privileges.
```

See https://github.com/opentofu/get.opentofu.org/blob/ddfdb1e4388db6f29285cace3d7e9cb77de188c7/src/install-opentofu.ps1#L564C49-L564C95

File: src/install-opentofu.ps1 (hunk 1 above)
Comment (pdecat) [reply src/install-opentofu.ps1:34]: My bad, comments in this script are *after* the parameter definition :upside_down_face:

File: src/install-opentofu.ps1 (hunk 2 above)
Comment (pdecat) [reply src/install-opentofu.ps1:35]: My bad, comments in this script are *after* the parameter definition :upside_down_face:

File: src/install-opentofu.ps1 [hunk 3]
```diff
@@ -290,8 +351,19 @@ function installStandalone() {
...
+        {
+            $dlFiles += $sigFile
+            $dlFiles += $certFile
+        } elseif ($verifyMethod -eq "gpg") {
+            if (@("1.6.0-alpha1","1.6.0-alpha2","1.6.0-alpha3","1.6.0-alpha4","1.6.0-alpha5","1.6.0-beta1","1.6.0-beta2","1.6.0-beta3","1.6.0-beta4","1.6.0-beta5","1.6.0-rc1").Contains($opentofuVersion)) {
+                throw [InstallRequirementNotMetException]::new("${bold}Additional tools are needed for the installation. Please read the following text carefully!${normal}`n`nOpenTofu version ${opentofuVersion} is not GPG-signed. This installer tries to verify that the OpenTofu version downloaded has been ${bold}signed by OpenTofu and has not been tampered with${normal}. This is only possible if either cosign or GPG is installed on the system. You have GPG installed, but only OpenTofu 1.6.0 and later are GPG-signed. You have the following options:`n`n1. ${bold}Install cosign${normal} and add it to your ${magenta}PATH${normal} or provide the ${magenta}-cosignPath${normal} parameter to your cosign installation.`n2. ${bold}Disable integrity verification${normal} with ${magenta}-skipVerify${normal} (${red}not recommended${normal}).")
```
Comment (pdecat) [inline src/install-opentofu.ps1:360]:
```suggestion
                throw [InstallRequirementNotMetException]::new("${bold}Additional tools are needed for the installation. Please read the following text carefully!${normal}`n`nOpenTofu version ${opentofuVersion} is not GPG-signed. This installer tries to verify that the OpenTofu version downloaded has been ${bold}signed by OpenTofu and has not been tampered with${normal}. This is only possible if either cosign or GPG is installed on the system. You have GPG installed, but only OpenTofu 1.6.0 and later are GPG-signed. You have the following options:`n`n1. ${bold}Install cosign${normal} and add it to your ${magenta}PATH${normal} or provide the ${magenta}-cosignPath${normal} parameter to your cosign installation.`n2. ${bold}Disabling integrity verification${normal} with ${magenta}-skipVerify${normal} (${red}not recommended${normal}).")
```

File: src/install-opentofu.sh [hunk 4]
```diff
@@ -712,10 +718,18 @@ install_standalone() {
...
+    if command_exists "${COSIGN_PATH}"; then
+      VERIFY_METHOD="cosign"
+    elif command_exists "${GPG_PATH}"; then
+      VERIFY_METHOD="gpg"
+    else
+      log_error "Additional tools are needed for the installation. Please read the following text carefully!"
+      log_error "${normal}This installer tries to verify that the OpenTofu version downloaded has been ${bold}signed by OpenTofu and has not been tampered with${normal}. This is only possible if either cosign or GPG is installed on the system, but neither was found. You have the following options:"
+      log_error "${normal}1. ${bold}Install cosign${normal} and add it to your ${magenta}PATH${normal} or provide the ${magenta}--cosign-path${normal} parameter to your cosign installation."
+      log_error "${normal}2. ${bold}Install GPG${normal} and add it to your ${magenta}PATH${normal} or provide the ${magenta}--gpg-path${normal} parameter to your GPG installation."
+      log_error "${normal}3. ${bold}Disable integrity verification${normal} with ${magenta}-skipVerify${normal} (${red}not recommended${normal})."
```
Comment (pdecat) [inline src/install-opentofu.sh:732]:
```suggestion
      log_error "${normal}3. ${bold}Disabling integrity verification${normal} with ${magenta}-skipVerify${normal} (${red}not recommended${normal})."
```

File: src/install-opentofu.ps1 (hunk 1 above)
Comment (eddie-knight) [reply src/install-opentofu.ps1:34]: @janosdebugs — you might consider putting a newline space between each param to reduce this confusion in the future

**Review by cam72cam — APPROVED** [review]

File: src/install-opentofu.ps1 (hunk 1 above)
Comment (janosdebugs) [reply src/install-opentofu.ps1]: We rolled back the Powershell changes as they have issues, they will follow in a later PR. Thank you everyone!

File: src/install-opentofu.ps1 (hunk 3 above)
Comment (janosdebugs) [reply src/install-opentofu.ps1]: Thank you for this, we are moving the Powershell changes to another PR, I'll add your changes there.

**Pull request merged**